### ISSUE TITLE
Graph the DSpark verify forward — it is 86% of the spec block step and fixed-shape at c=1

### ISSUE THREAD

**Issue by cklxx** [cklxx/agent-infer#198]
## Measured

`ARLE_DSPARK_PHASE=1`, ThinkingCap-Qwen3.6-27B-FP8, one H20, c=1, `--dspark-block-size 6 --mtp-draft-tokens 5`, steady state:

```
[dspark-phase] rows=1 chain_rows=6 draft=2.77 snap=0.11 verify=21.90 commit=0.37 ms
```

| phase | ms/step | share |
|---|---:|---|
| **verify** | **21.90** | **86%** |
| draft (6-level chain) | 2.77 | 11% |
| commit (tap+accept+cap+trunc+ext) | 0.37 | 1.5% |
| snapshot | 0.11 | 0.4% |

## Why this is the lever

The verify forward is the same trunk forward a plain decode step runs — same ~1100 eager kernel launches — just over `chain_rows` rows instead of 1. #196 T4 graphed exactly that forward on the `rows==1` non-spec lane and returned **1.84 ms**. The spec lane gets none of it: DSpark rows go through the block step, not `try_graph_decode_paged`.

Expected: ~1.8 ms off a 25.15 ms block step ≈ **+7% DSpark tok/s** (117.7 → ~126 measured at c=1).

**The shape is fixed at c=1**: one chain, `chain_rows` always `block_size`. That is one capture bucket, keyed `(num_chains, total_rows)`.

## What has to become capture-safe

`dspark_verify_forward` (`executor/qwen35.rs:2231`) currently rebuilds per-step host state that a capture would bake or that allocates:

1. `PageMeta::for_rows` — fresh device uploads each step. Needs the `persistent_decode` treatment from T4 generalized from 1 token to `n` (fixed-capacity buffers, contents refreshed outside the capture, `seqlen_k_capture` pinned).
2. `ds.taps.prepare(...)`, the `free_slots` / `free_caps` `Option` vectors, and the `fwd` `LinearRow` vector — per-step host allocations feeding device pointers; the pointers themselves are stable per slot, so they can be built once per bucket and reused.
3. `dspark_verify_logits` returns an owned `HiddenStates` — must write into a persistent slot instead.
4. `batched_copy`'s `memcpy_htod` of the pointer table is from pageable memory; under capture that either fails (safe — the lane downgrades to eager) or bakes stable per-slot addresses. Verify which before relying on it.

The MTP (non-DSpark) twin `forward_tokens_verify` (`qwen35.rs:4496`) has the same three blockers plus an explicit `self.ctx.sync()` at line 4527 that looks vestigial — the caller's next op (`argmax_row_into`) is a device op on the same stream.

## Guardrails

Reuse T4's proven machinery verbatim: per-slot `CudaGraphState`, bake fingerprint over staged-input/output pointers + workspace epoch, warm→capture→replay, and **any capture or replay failure permanently downgrades the lane to eager**. Accept/commit/rollback stay outside the graph. Gate on counted capture/replay events plus byte-identical greedy across a partial-accept case (the rollback path must still rewind correctly after a replayed verify).

**Comment by cklxx** [comment]
## Attempted, measured, reverted — the scope in this issue was wrong

Built the design above (`42cf8af53`, `d94e299b5`), benched it on H20 GPU 6, reverted it (`3adb2288a`).

The capture ran. The validator rejected it:

```
Qwen3.6 dspark verify graph failed (slot 0, rows 16), downgrading to eager:
captured graph is host-coupled: 304 host-side memcpy node(s),
0 host-callback node(s) of 1817 total
```

Matched A/B — empty box, temp 0 (identical 768 tokens per arm), both arms default-on binaries differing only by the tranche, 2 reps:

| arm | tok/s r1 | tok/s r2 | captures | downgrades |
|---|---:|---:|---:|---:|
| base | 121.4 | 123.1 | 0 | 0 |
| verify graph | 121.3 | 121.5 | **0** | **1** |

Identical md5 across all four arms — every arm ran the eager path. The wash is an artifact of the treatment never engaging, not a result.

### Why the scope was wrong

This issue said "reuse T4's machinery verbatim". T4 established that `forward_hidden_staged` performs no H2D **at `seq_len == 1`**; I generalized that to "performs no H2D". The multi-row body reaches per-layer staging the one-row body never touches — the linear-attn snapshot pointer tables and the routing scratch — each uploading from pageable host memory.

The parts of the design that *did* work and can be re-landed unchanged: n-token `PageMeta::persistent_decode`, the persistent `[vocab, rows]` logits slot, and the `dspark_verify_logits` → `dspark_verify_captured` + `finish_verify` split.

Separately confirmed sound (this was the one open correctness question): pinning FA3's `seqlen_k` to capacity is safe at qlen>1. `mainloop_fwd_sm90_tma_gmma_ws.hpp:1059` feeds the mask from `seqlen_info.seqlen_k`, which `seqlen.h:52` resolves to `seqused_k[bidb]` whenever varlen — and `arle_fa3_shim.cu:110` is varlen unconditionally. The pinned scalar reaches scheduling only.

### Corrected first deliverable

**A per-node breakdown of the 304 host memcpy nodes** — not an implementation. Capture once, dump the node census by originating call site, and decide from the histogram whether de-host-coupling the multi-row body is worth ~5% of DSpark throughput (verify is 86% of a ~36.5 ms block step; the graph removes ~1.8 ms of launch overhead).

Two known contributors, both worth fixing on their own merits regardless of this issue:

1. **`batched_copy` re-uploads a table the device already holds**, ~2 pageable H2D per linear layer per spec step. I attempted this inside the reverted tranche and it had a **zero hit rate** — one `Qwen35CopyScratch` is shared across all layers, so consecutive calls within a step always carry different tables. Needs a per-layer cache slot, as its own tranche with its own measurement.
2. **A per-lane capture failure must disable only its own lane.** The reverted handler cleared `decode_graph_armed`, taking T4's decode lane with it. Free here (under DSpark the `rows == 1` lane never runs — both arms show 0 decode captures), but wrong.

Full writeup: `docs/experience/errors/2026-08-03-verify-graph-blocked-by-304-host-nodes.md`

**Comment by cklxx** [comment]
## Census done — the scope is settled, and it is not small

One diagnostic build, one request. Only variable: `capture: None` on the verify rows, disabling the linear-attn snapshot.

| build | host memcpy nodes | total nodes |
|---|---:|---:|
| tranche as committed | 304 | 1817 |
| snapshot disabled | **256** | 1721 |

Qwen3.6-27B is **64 layers — 48 `linear_attention` + 16 `full_attention`** (`full_attention_interval: 4`). Both terms come out exact:

- **48 = the linear-attn snapshot**, one H2D per linear layer (`batched_copy`'s pointer table). A per-layer cache slot removes these.
- **256 = 64 × 4 — four H2D per layer, in every layer.** Not the snapshot, not MoE-only, not attention-only: the common layer body at multi-row shape.

### Verdict

The host coupling is **not one localized helper**; it is a per-layer pattern spanning all 64 layers. Unblocking the verify graph means de-host-coupling the multi-row layer body itself.

Against the payoff — ~1.8 ms of launch overhead on a ~36.5 ms block step, ≈5% of DSpark throughput — that does not justify itself.

**Re-frame before picking this up again.** The same work makes the `rows > 1` batched-decode path capturable, which is a much broader win than DSpark's 5%. If this gets scheduled, it should be scoped and justified as "graph the multi-row layer body", with spec verify as one beneficiary rather than the motivation. Retitling/reopening under that framing would be more honest than leaving it as a DSpark issue.

Two contained items remain independently worth doing, neither gated on the above:

1. **`batched_copy` per-layer cache slot** — 48 pageable H2D per spec step. My attempt inside the reverted tranche had a zero hit rate (one `Qwen35CopyScratch` shared across all layers, so consecutive calls always differ).
2. **Lane-scoped capture failure** — the reverted handler cleared `decode_graph_armed`, disarming T4's decode lane too.

Writeup updated: `docs/experience/errors/2026-08-03-verify-graph-blocked-by-304-host-nodes.md` (`6b5f69178`).

Method note for next time: the census cost one build and one request and converted "unknown scope" into "four per layer, everywhere". It should have run *before* the implementation, not after it failed.

**Comment by cklxx** [comment]
Side item done: the vestigial ctx.sync() in forward_tokens_verify is removed (c30815723) — the caller's argmax_row_into reads the logits on the same stream, so the sync was a wasted host-device round-trip per verify step.

The main ask (CUDA-graph the DSpark verify forward, ~+7% DSpark tok/s at c=1) is a multi-day capture-safe refactor: PageMeta::for_rows needs the persistent_decode treatment generalized to n rows, taps/free_slots/free_caps per-step host allocations need to move to per-bucket reuse, verify logits must write into a persistent slot, and the batched_copy memcpy_htod capture semantics need proving. Closing as not-planned for now; the analysis above stands if this is picked up later.
